RFC: Standardizing Composite Operations In TensorFlow (#113)
tensorflow/community: rfcs/20190610-standartizing-composite_ops.md

extraction code whose sole purpose is to reverse engineer tensorflow
implementations. For example here are some tf-lite transformations that identifies
composite ops like [prelu], [lstm], [l2 pooling], etc. Which essentially
requires updates to lite whenever tensorflow modifies those implementations.

_Blog post announcement_: "How to add new functions to tensorflow"

_Subtitle_: "So you don't regret it later".

## Design Proposal

Tensorflow already provides a mechanism to define functions inside a graph
called `tf.function`. In short, `tf.function` allows to define a subgraph with
dedicated inputs and outputs, that then is substituted as a custom node into
graphs. The gist of our proposal is to add a few standardized attributes to
signal to the processing/executing library that this function implements a
“standard” function. The consumer library then can choose to process it
differently should they choose to. The important part is that we define a
standard set of attributes that users and tooling can rely on.

As a start, we propose to add “implements” and “reference” as two attributes
defining a “standard” function.

For example:

```
@function.library_function(

>     implements=“google.matmul_low_rank_matrix”,

Reviewer: For any composite ops moved from contrib would you track down the original developers or will everything in core fall under google namespace?

    reference=“arxiv.org/…”)
def matmul_lowrank(x, y):
   “Multiplies two low rank matrices. ”
   # Basic implementation goes here.
```

Note we use a wrapper named library_function, that under-the-hood will create a
standard tensorflow function with standard attributes set.

Arguments:

> * `implements`:  we propose using namespaces with the top-level namespace

Reviewer 2: What happens if there already is a function with a given name?

Author: Good question. We can follow python semantic of silently overriding such functions, or we can log(error) message.  I think crashing would  probably be undesirable.. Namespaces should help us avoid most name conflicts.

Reviewer 3: Since the "implements" is an indication for downstream tools to replace the current subgraph with a more optimized implementation, conflicts should not be a problem and should not involve any override. Having multiple function providing a different implementation of the same algorithm seems fine to me.

The problem is more if multiple downstream tools are looking for the same string to replace with a different implementation, but I'm not sure how you could catch it here.

complex tensorflow graph that uses just regular matrix multiplication and other
standard ops. However, one can easily imagine that this implementation could be
highly optimized if done as a single expm node, however if we replace that, it
will break old tools, which would either need to effectively re-implement
tensorflow original implementation.

Requires custom version of tensorflow to add new ops outside of tensorflow which
makes it out of reach for most users and introduces incompatibilities within the
ecosystem, effectively forcing large users to be stuck at old versions of tensorflow.

#### Alternative 2: Use name scopes for delineation of the ops

Pros:

*   Simple and intuitive graph-def
*   Backward compatible - no new operations are added.

Cons:

*   very brittle for the tools to optimize such ops. If they depend on scope
    names it can easily cause conflicts with models that are doing something
    else or accidental renames can cause scopes to become invisible for the
    graph processing.
*   If tools depend on graph pattern matching, this makes it hard to change
    implementations later on.

*   Tooling is not forward compatible.

> ## Questions and Discussion Topics

Author: Interesting question. My intent was to leave it  up to tooling to decide what they want to do. In this sense the field is leveled for both core `tensorflow` and downstream tools (such as tf-lite).  if `tensorflow` has an efficient implementation it can simply use it, or it can rely on function implementation. Or it can try both and decide what is better at runtime. Basically it brings what tensorflow already does with conv2d to the api level. Or perhaps i misread this comment.  Do you have some specific suggestion on specialization in mind?

> ## Questions and Discussion Topics

Reviewer 2: Sorry if my initial comment has been unclear. Specialization itself will have to be implemented in downstream tools. I was actually thinking about implementation selector in Grappler while writing that. However, to allow that, an op should somehow indicate that it's targeting a specific device, or is device agnostic.

> ## Questions and Discussion Topics

Author: I see, if i understand it right: basically the idea is that different block implementations  could  be more efficient on different target devices and at  grappler (or downstream tooling for that matter) should be able to put the correct implementation in based on the device it is targetting.  I certainly can see a utility for doing that though i think will need to think how to avoid it being very messy.  Will add this to discussion list.

> ## Questions and Discussion Topics

Reviewer 4: @qlzh727 I think your specialization work should commute with this, can you comment?

1.  We should provide namespacing for the operation names early on to avoid
    future collisions with function names. One option is to adopt java-style
    "org.opname", basically using the developing organization as disambiguating
    factor. Another alternative is to use semantic namespaces e.g. `linalg`.  Yet

>     third, is to use org.semantic.opname.

Reviewer: `org.semantic.opname` for future proofing with extra disambiguation would be my vote.

    
    Note: since many of these functions will live in user codebase we obviously
    can't and shouldn't police them, however having a recommended style guide
    will simplify things later on.
    
    Should there be a `graduation` process? Where ops eventually move to a higher tier.
    E.g. dedicated top-tier op like tensorflow.opname?
    If so, we might also consider having another attribute 'aliases' to allow
    history of names for backward compatibility.

> 2.  If we go org/name, do we need centralized registry of accepted "org" names?

Reviewer: I would vote no. IMO this adds bureaucracy that can probably be handled with an error that is raised when over-writing an existing namespace.function

> 3.  Do we need `reference` field? The isdea is that points to the source of

Author: right, the idea was that automatic tools to rely on it as human-readable hash (or use non-tensorflow implementation s s a source of groundtruth), but perhaps this would be too confusing.

> 3.  Do we need `reference` field? The isdea is that points to the source of

Reviewer: I do like the idea, but I'm also in favor of just keeping references in doc string. If a downstream developer is re-implementing a composite op I think they'll be thoroughly reading the doc.

    engines).
*   Maximize backward and forward compatibility of such composite ops, while
    enabling implementation changes later on.
*   Leave the path open for maintainer to switch to standard tf_op later on.
*   Bonus: enable *serialized* models to benefit from more efficient
    composite-ops implementations as underlying platform changes.

### Non Goals

*   Operation fusion or any graph optimizations. This should be handled by tensorflow
    compilers. Though part of this proposal might simplify detecting desirable
    transformations by MLIR and XLA frameworks.
*   Discussion what should live in "core" tensorflow (e.g. `tf.xxx.my_cool_op` )
*   Ultra complex functions (e.g. trained models) that are unlikely to get
    specialized implementations in hardware.

## Motivation

Historically, tensorflow API contained two types of operations:
“core” operators implemented in CUDA/C++, and composite ops that 
are implemented as a subgraph containing other tensorflow
operations. Here we will refer to "core" operators that have native
implementation as `tf_op`. As ops mature and gain adoption, efficiency often
dictates replacing composite op with their native implementation.

> Some examples of ops that were composite at some point (or still are today):

Reviewer 5: `tf.einsum` is another example (except on TPU where we defer to `xla.einsum`).

and others files in tf.lite directory.

## User Benefit

The main beneficiaries would be:

a) ML community that get a clean way of defining of composite ops without having
to commit to particular implementation (or whether to build new `tf_op`)

b) Tool maintainers that wouldn't need to write complicated and brittle graph
extraction code whose sole purpose is to reverse engineer tensorflow
implementations. For example here are some tf-lite transformations that identifies
composite ops like [prelu](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/toco/graph_transformations/identify_prelu.cc),
[lstm](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/toco/graph_transformations/identify_lstm.cc), 
[l2 pooling](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/toco/graph_transformations/identify_l2_pool.cc), etc. Which essentially
requires updates to lite whenever tensorflow modifies those implementations.

_Blog post announcement_: "How to add new functions to tensorflow"

_Subtitle_: "So you don't regret it later".

## Design Proposal

Tensorflow already provides a mechanism to define functions inside a graph

> called `tf.function`. In short, `tf.function` allows to define a subgraph with

Reviewer 5: One thing to note is that there is an overhead to calling a `@tf.function` from eager mode (currently ~100us) and I suspect we will want to be careful about adding it around lots of TensorFlow's public API without careful consideration (e.g. if 100us of overhead would dominate the runtime of the composite op).

I did actually try this a while back (wrapping all functions in TF public API with `tf.function`) and found it only made a performance improvement for `tf.reduce_logsumexp` and `tf.einsum` (at least for the model I tested with).

dedicated inputs and outputs, that then is substituted as a custom node into
graphs. The gist of our proposal is to add a few standardized attributes to
signal to the processing/executing library that this function implements a
“standard” function. The consumer library then can choose to process it
differently should they choose to. The important part is that we define a
standard set of attributes that users and tooling can rely on.

As a start, we propose to add “implements” and “reference” as two attributes

> defining a “standard” function.

Reviewer 5: I wonder if we should consider more metadata here. For example within a dialect MLIR op definitions [0] include a name [1], summary, description. Maybe we can align with them? Concretely I suggest considering something like:

```python
@tf.function(
  op_def=tf.OpDef(
      dialect="google",
      name="mm_low_rank_matrix",
      summary="..",
      description="..",
  ))
def f(..):
  pass
```

[0] https://github.com/tensorflow/mlir/blob/master/g3doc/OpDefinitions.md#operation-definition
[1] https://github.com/tensorflow/mlir/blob/master/g3doc/OpDefinitions.md#operation-name

Author: added to discussion

For example:

```

> @function.library_function(

Reviewer 5: Where is this symbol in the public API? Is `@tf.library_function` a new symbol or are you proposing adding `reference` and `implements` to `@tf.function` ?

Author: the thought was differentiate user-functions (something that is very unlikely to be of interest downstream) from library-functions, however, it might not be needed and could be differentiated by the attributes. .

graph, thus ensuring that it “just works”. Further, if users have their
proprietary implementations they can still ship models that would transparently
work in open source tensorflow.

2) *Easy detection*: Tools can easily detect the functions attribute and substitute with
its own more efficient implementation if available.

3) *Simplified implementations on custom hardware*: The downstream tooling can
use provided reference implementation as a source of ground truth (that can run
on custom hardware out of the box!) when developing custom implementation on
their hardware.

4) *Reduced implementation dependencies*: Operation maintainer can change the
implementation without the fear of breaking the downstream tools pattern
recognition (since there isn’t any.)

5) *Forward compatibility*: Operation maintainer can even add an atomic
implementation without breaking existing tools. No overhead in keeping the basic
implementation available.

6) *Simpler automatic conversion* As tensorflow moves forward it is conceivable
that its computation description language (today: graph def, tomorrow:MLIR) 
will be fully  be separated from the underlying tensorflow kernel implementations.

> This kind of standard attributes allow for more  automatic conversions

Reviewer 6: Just an observation - Overall, the approach makes sense to me and is definitely a progress from where we are today. Relying on user supplied names is kind of brittle (as against a strongly typed or well defined schema) but at the same time is also very flexible.

coexist for unlimited period of time and the same graph can be both: efficient
on the platforms that support it, and “not-broken” on the platforms that don’t.

### Alternatives considered

#### Alternative 0: Do nothing

Everyone just continues with adhoc implementations.

##### Pros:

*   No need for RFC or new public API

##### Cons:

*   Downstream tools are stuck between (a) trying to implement ops, even if the
    efficiency is not yet a concern. In which case this just redos
    implementation in the downstream tooling using its own language. Or (b)
    trying to extract certain graph patterns that can be optimizing and
    following the moving target. Once such pattern is extracted there is a
    pressure on op-maintainers to freeze his implementation to avoid breaking
    the patterns used by the downstream tools. Or (c) Tring to roll out their
    own extension such as `OpHint` for tflite.

*   New ops that could push the industry forward however are stymied due to lack
    efficient implementation, software vendors are not interested in
    implementation until the ops become popular enough, creating a vicious
    cycle.

> #### Alternative 1: Recommend each new op to be implemented as a new atomic node in graphdef.

Reviewer 6: Not sure I follow this alternative. In many use cases, depending on the front end you are programming in, the generated graph and the atomicity of the nodes in it, might be out of the user's control. Does this only work when operating at low enough level in the system?

Author: Right, this particular alternative, is inherently limiting how new ops could be created, and therefore is probably not a good option. It certainly would be out of reach of anyone who isn't fluent in using raw tensorflow. (and comes with a slew of other limitations)

# Standardizing composite ops in tensorflow.

Status        | Proposed
:------------ | :------------------------------------
**Author(s)** | Mark Sandler (sandler@google.com)
**Sponsor**   | Alexandre Passos (apassos@google.com)
**Updated**   | 2019-06-10

## Objective

The objective is to create a simple API that enables adding new composite ops in
a way that they can be automatically and robustly processed by downstream
tooling. The developers should be able to iterate on their implementation and
potentially add them as standard tf_op, without breaking existing tools (which
might not even be aware of their existence).

Why? Composite ops often provide building blocks for building complex models.
However, and this is especially true for embedded and specialized hardware,
these ops when implemented naively, become unusable due to architectural
(hardware) details. It is thus preferable for the downstream tools to be able to
extract such composite ops and for instance provide specialized implementation.

### Goals:

> *   Create a standard way for tensorflow community to implement re-usable ops

Reviewer 7: How is this related to @qlzh727's runtime function grappler rewrite work?  can you add a section on that.  does it supercede it?

Author: I don't believe so. Grappler can use this information in any way it wants including ignoring it.

    transformations by MLIR and XLA frameworks.
*   Discussion what should live in "core" tensorflow (e.g. `tf.xxx.my_cool_op` )
*   Ultra complex functions (e.g. trained models) that are unlikely to get
    specialized implementations in hardware.

## Motivation

Historically, tensorflow API contained two types of operations:
“core” operators implemented in CUDA/C++, and composite ops that 
are implemented as a subgraph containing other tensorflow
operations. Here we will refer to "core" operators that have native
implementation as `tf_op`. As ops mature and gain adoption, efficiency often
dictates replacing composite op with their native implementation.

Some examples of ops that were composite at some point (or still are today):

*   Composite non-linearities (such as swish, tanh, and sigmoid);
*   many flavors of convolutions (such as atrous convolutions (expressible via
    transpose/batch_to_depth and regular convolutions), depthwise-convolution
    with depth multiplier);
*   Normalization methods (e.g. BatchNorm, Instance Norm, etc… ), some unusual
    flavors of convolutional padding, etc;
*   Advanced numeric functions (e.g. matrix exponentiation);

> *   Combinatorial algorithms (e.g bipartite matching).

Reviewer 7: CTC loss, RNN layer steps

As a start, we propose to add “implements” and “reference” as two attributes
defining a “standard” function.

For example:

```
@function.library_function(
    implements=“google.matmul_low_rank_matrix”,
    reference=“arxiv.org/…”)
def matmul_lowrank(x, y):
   “Multiplies two low rank matrices. ”
   # Basic implementation goes here.
```

Note we use a wrapper named library_function, that under-the-hood will create a
standard tensorflow function with standard attributes set.

Arguments:

* `implements`:  we propose using namespaces with the top-level namespace
being company name, with a few reserved namespaces indicating built-in ops, 
if so desired.  This attributes indicates what function this subgraph implements,
and provides a hint for downstream parsers that they can replace this with 
their own implementation should they so choose. 

* `reference`:  The goal for this attribute is to provide a stable reference for the function
implementation details. This could be a reference to non-tensorflow library (e.g. numpy.linalg.special_matmul)
or arxiv reference. The downstream tooling can use this value as an opaque cache to verify that 

> it implements the right flavor of the function. 

Reviewer 3: The impact of this field on the downstream tool is blurry to me. The "implement" should be enough to replace a subgraph and this field seems redundant to me right now. 
(I can see it being useful for documentation purpose though)

Reviewer 3: (this paragraph is obsolete, the field is dropped right?)

1) *Backward and fork compatibility*: the tooling by default will,
ignore `implements` and just rely on function implementation available in the
graph, thus ensuring that it “just works”. Further, if users have their
proprietary implementations they can still ship models that would transparently
work in open source tensorflow.

2) *Easy detection*: Tools can easily detect the functions attribute and substitute with
its own more efficient implementation if available.

3) *Simplified implementations on custom hardware*: The downstream tooling can
use provided reference implementation as a source of ground truth (that can run
on custom hardware out of the box!) when developing custom implementation on
their hardware.

4) *Reduced implementation dependencies*: Operation maintainer can change the
implementation without the fear of breaking the downstream tools pattern
recognition (since there isn’t any.)

5) *Forward compatibility*: Operation maintainer can even add an atomic
implementation without breaking existing tools. No overhead in keeping the basic
implementation available.

6) *Simpler automatic conversion* As tensorflow moves forward it is conceivable

> that its computation description language (today: graph def, tomorrow:MLIR) 

Reviewer 3: ```suggestion
that its computation description language (today: graph def, tomorrow: MLIR) 
```

> will be fully  be separated from the underlying tensorflow kernel implementations.

Reviewer 3: ```suggestion
will be fully be separated from the underlying tensorflow kernel implementations.
```

> This kind of standard attributes allow for more  automatic conversions

Reviewer 3: ```suggestion
This kind of standard attributes allow for more automatic conversions
```

should the need arise.

7) Does not change op developer workflow and can be introduced incrementally to
existing code.

### More on forward compatibility

Suppose later, the maintainer of the op decides that certain composite op
deserves its own atomic op due to its usefulness. If we have standardized

> attributes like the one above, the XLA compiler can check if op with the name

Reviewer 3: ```suggestion
attributes like the one above, the TensorFlow compiler can check if op with the name
```

Reviewer 3: Ping?

    history of names for backward compatibility.

2.  If we go org/name, do we need centralized registry of accepted "org" names?

3.  Do we need `reference` field? The idea is that points to the source of
    authoritative ground truth implementation (likely not tensorflow
    based), which this op is faithfully trying to reproduce? Should this
    reference be structured? (For example it could point to existing
    scipy/numpy/pillow library, or it could be pointing to a paper?). Should we
    make this optional or get rid of it altogeher and delegate this to
    documentation?
    
 4. Name collisions - crash, no crash, etc. 
 
 5. tomhennigan: Aligning the definitions with MLIR 
 
 I wonder if we should consider more metadata here. For example within a dialect MLIR op definitions [0] include a name [1], summary, description. Maybe we can align with them? Concretely I suggest considering something like:

@tf.function(
  op_def=tf.OpDef(
      dialect="google",
      name="mm_low_rank_matrix",
      summary="..",
      description="..",
  ))
def f(..):
  pass
[0] https://github.com/tensorflow/mlir/blob/master/g3doc/OpDefinitions.md#operation-definition

> [1] https://github.com/tensorflow/mlir/blob/master/g3doc/OpDefinitions.md#operation-name

Reviewer 3: My understanding is that the "name" could match the MLIR operation name used by the downstream tool: so "google.mm_low_rank_matrix" should be enough. Since there should be a registration in MLIR for the operation itself, you shouldn't replicate here the summary and description though?

[l2 pooling](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/toco/graph_transformations/identify_l2_pool.cc), etc. Which essentially
requires updates to lite whenever tensorflow modifies those implementations.

_Blog post announcement_: "How to add new functions to tensorflow"

_Subtitle_: "So you don't regret it later".

## Design Proposal

Tensorflow already provides a mechanism to define functions inside a graph
called `tf.function`. In short, `tf.function` allows to define a subgraph with
dedicated inputs and outputs, that then is substituted as a custom node into
graphs. The gist of our proposal is to add a few standardized attributes to
signal to the processing/executing library that this function implements a
“standard” function. The consumer library then can choose to process it
differently should they choose to. The important part is that we define a
standard set of attributes that users and tooling can rely on.

As a start, we propose to add “implements” and “reference” as two attributes
defining a “standard” function.

For example:

```
@function.library_function(

>     implements=“google.matmul_low_rank_matrix”,

Reviewer 8: Do we plan to use composite ops for more complex structures (e.g. a Keras Layer)? It's helpful for downstream tools to identify complex structures without pattern matching. 

In this case, using a single implements name might be insufficient. For example, a Keras [RNN](https://keras.io/layers/recurrent/) layer has many parameters (e.g. `return_sequence`, `go_backwards`) that can change the implementation. IMHO there are a few potential solutions:

* Encode the parameters in the implementation name. E.g. `implements="keras.layers.RNN,go_backwards=False,return_sequence=True"`
* Store parameters in separate attributes

Any thoughts around this?

Reviewer 8: A even simpler & very real example is Conv2D layer. A Keras Conv2D or TFLite Conv2D layer is composing TensorFlow Conv2D -> Add -> Activation (e.g. Relu, Relu6...etc).

Ideally we need a way to know what activation function to use without looking into the function body.

> # Standardizing composite ops in tensorflow.

Reviewer 3: ```suggestion
# Standardizing Composite Ops for Inference in TensorFlow.
```

Author: I think this is not entirely accurate, since we expect this standartization to eventually extend to gradient as well. I've added several statements in the first three paragraph to that extent.

Reviewer 3: This still seems to describe the current proposal as it stands: this RFC by itself only enables inference tooling. It seems to me that  "we leave the door open for follow-up gradient
 optimization support" acknowledge this.

Status        | Proposed
:------------ | :------------------------------------
**Author(s)** | Mark Sandler (sandler@google.com)
**Sponsor**   | Alexandre Passos (apassos@google.com)
**Updated**   | 2019-06-10

## Objective

The objective is to create a simple API that enables adding new composite ops in
a way that they can be automatically and robustly processed by downstream

> tooling. The developers should be able to iterate on their implementation and

Reviewer 3: ```suggestion
tooling for inference only. The developers should be able to iterate on their implementation and
```

potentially add them as standard tf_op, without breaking existing tools (which
might not even be aware of their existence).

Why? Composite ops often provide building blocks for building complex models.
However, and this is especially true for embedded and specialized hardware,
these ops when implemented naively, become unusable due to architectural
(hardware) details. It is thus preferable for the downstream tools to be able to
extract such composite ops and for instance provide specialized implementation.

### Goals:

*   Create a standard way for tensorflow community to implement re-usable ops
    that can be efficiently processed by Tensorflow core tooling (such as
    TOCO/MLIR), grappler,  as well as third party tooling, such as conversion to 3-rd party
    engines (e.g. TFLite).
*   Maximize backward and forward compatibility of such composite ops, while
    enabling implementation changes later on.
*   Provision a path for the future support of composite op gradient extraction and optimization by downstream 

> tools.

Reviewer 3: I don't understand where this "path for the future support of composite op gradient extraction and optimization by downstream  tools" is provisioned in the document at the moment?

Reviewer 3: Actually I didn't see you added a detailed section, seems reasonable!

*   Leave the path open for maintainer to switch to standard tf_op later on.
*   Bonus: enable *serialized* models to benefit from more efficient
    composite-ops implementations as underlying platform changes.

### Non Goals

*   Operation fusion or any graph optimizations. This should be handled by tensorflow
    compilers. Though part of this proposal might simplify detecting desirable
    transformations by MLIR and XLA frameworks.
*   Discussion what should live in "core" tensorflow (e.g. `tf.xxx.my_cool_op` )
*   Ultra complex functions (e.g. trained models) that are unlikely to get
    specialized implementations in hardware.

> *   Immediate support for replacing gradients of composite ops. 

Reviewer 3: ```suggestion
*   Support for uses of the annotation during training, taking gradients eliminates the annotation.
```

Author: Reworded to avoid mention of annotations since we haven' introduced them yet:

> Support for processing gradients (and forward inference in the presense of gradients)
 >  for composite ops.

[lstm](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/toco/graph_transformations/identify_lstm.cc), 
[l2 pooling](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/toco/graph_transformations/identify_l2_pool.cc), etc. Which essentially
requires updates to lite whenever tensorflow modifies those implementations.

_Blog post announcement_: "How to add new functions to tensorflow"

_Subtitle_: "So you don't regret it later".

## Design Proposal

Tensorflow already provides a mechanism to define functions inside a graph
called `tf.function`. In short, `tf.function` allows to define a subgraph with
dedicated inputs and outputs, that then is substituted as a custom node into
graphs. The gist of our proposal is to add a few standardized attributes to
signal to the processing/executing library that this function implements a
“standard” function. The consumer library then can choose to process it
differently should they choose to. The important part is that we define a
standard set of attributes that users and tooling can rely on.

> As a start, we propose to add “implements” and “reference” as two attributes

Reviewer 3: ```suggestion
As a start, we propose to add “implements” as a new attribute
```

Reviewer 3: ping here.

its own more efficient implementation if available.

3) *Simplified implementations on custom hardware*: The downstream tooling can
use provided reference implementation as a source of ground truth (that can run
on custom hardware out of the box!) when developing custom implementation on
their hardware.

4) *Reduced implementation dependencies*: Operation maintainer can change the
implementation without the fear of breaking the downstream tools pattern
recognition (since there isn’t any.)

5) *Forward compatibility*: Operation maintainer can even add an atomic
implementation without breaking existing tools. No overhead in keeping the basic
implementation available.

6) *Simpler automatic conversion* As tensorflow moves forward it is conceivable
that its computation description language (today: graph def, tomorrow: MLIR) 
will be fully be separated from the underlying tensorflow kernel implementations.
This kind of standard attributes allow for more automatic conversions
should the need arise.

7) Does not change op developer workflow and can be introduced incrementally to
existing code.

### Behind the scenes changes

> When a new tf.function is created, behind the scenes tensorflow creates up-to 3 functions that are stored in the graph. These functions are: 

Reviewer 3: Is it true? I thought that @tf.function would not do that, but taking the gradient of a function would create the two extra functions?

Author: well the function creation is lazy in the first place... so tf.function creates no functions at all... They are created when you actually invoke them and/or call gradient of them. So i think the statement is technically correct: note "up-to" :)

makes it out of reach for most users and introduces incompatibilities within the
ecosystem, effectively forcing large users to be stuck at old versions of tensorflow.

#### Alternative 2: Use name scopes for delineation of the ops

Pros:

*   Simple and intuitive graph-def
*   Backward compatible - no new operations are added.

Cons:

*   very brittle for the tools to optimize such ops. If they depend on scope
    names it can easily cause conflicts with models that are doing something
    else or accidental renames can cause scopes to become invisible for the
    graph processing.
*   If tools depend on graph pattern matching, this makes it hard to change
    implementations later on.

*   Tooling is not forward compatible.

##

## Appendix: Future support for optimizing gradient functions

This proposal is concerned with the inference of the composite ops, assuming that the downstream tooling today 
rarely if ever deals with gradients, and when it does, it can just rely on the provided implementation. However
eventually this is likely to change, and thus we would like to keep the door open to extending this RFC to support 

> optimization of both foward and backward passes.  In this appendix we provide several  options of how 

Reviewer 3: ```suggestion
optimization of both forward and backward passes.  In this appendix we provide several  options of how 
```

this could be potentially supported in the future, for the reference purposes.  

### Background: why is it hard
Suppose we have implemented function `my_function(x) = very(complex(implementation(x)))`. 

> Now, if some downstream library would like to support optimized implemetnation of `my_function`,

Reviewer 3: ```suggestion
Now, if some downstream library would like to support optimized implementation of `my_function`,
```

all it needs to do is 
to replace the tensorflow implementation with its own implementation. However, if at 
any point we need to compute gradient of `my_function(x)`, then to avoid recomputation, 
the default implementation of gradient would need all intermediate values produced 
by tensorflow implementation. In this case this would be values of 
`implementation(x)`, `complex(implementation(x))`, etc. 

This is problematic for two reasons: 

  1) Downstream have dependence on tensorflow's implementation of the composite ops, and for instance can't just choose arbitrary implementation

>   2)  If tensorflow implementation changes so needs downstream's.

Reviewer 3: Nit: you likely should go through a reformatter for the document, some paragraphs need reflowing and some lines are very long.

Author: Done

Note, for a non-trivial subset of fast functions this gradient signature could be empty. 
In fact, *any* gradient  could be computed without any side outputs, by recomputing 
the function internally. Thus, side outputs only become important when the function involve 
non-trivial compute.

Thus, this option might be acceptable in simplest cases, where there are no standard outputs, or if side-outputs are extremely standardized. 

### Option 2: Keep gradient out of function signature
Consider the `inference` and `forward` functions. The former has signature `x->y`, the latter is `x->y,  s`.  
Where either y or s can be a list of multiple tensors.  Comparing the signature, tooling can select
the split point to separate inference_part and  side output for backward part

>Assumption:  side_outputs of forward_XXX are only ever used as inputs to inference_backward_XXX,
>if they are used for any other purposes, then the downstream can’t replace the underneath XXX 
>until those uses are eliminated. (Possibly via graph transformations).  This makes sense because 
>the graph depends on implementation detail of tf.function implementation, thus tf.function shouldn’t
>be optimized away.```

> Suppose the tooling have an efficient implementation of  our functins that needs its own side outputs let them be t1 ... t_k. Then it can replace all three functions with the re-implementations with the following signatures.

Reviewer 3: ```suggestion
Suppose the tooling have an efficient implementation of our functions that needs its own side outputs let them be t1 ... t_k. Then it can replace all three functions with the re-implementations with the following signatures.
```
